src/channel/mpmc.rs

```diff
@@ -182,12 +196,27 @@ where
         }
     }
 
+    /// Tries to read the value from the channel without blocking.
+    fn try_receive(&mut self) -> Option<T> {
+        if !self.buffer.is_empty() {
+            let val = self.buffer.pop();
+
+            // Since this means a space in the buffer had been freed,
+            // try to copy a value from a potential waiter into the channel.
+            unsafe { self.try_copy_value_from_last_waiter() };
+
+            Some(val)
+        } else {
+            None
```

Comment:
I think checking for a closed channel is also missing.

Reply:
Edit: Seems like this simply returns `None` for a closed channel and no value is available. Could return `Option<Option<T>>` to differentiate those. What do the other implementations do?

Reply:
Tokio doesn't implement try_recv yet. I think returning a result with `TryReceiveError` here would be better.

```rust
enum TryReceiveError {
  WouldBlock,
  Closed,
}
```

Reply:
That looks indeed better than returning a nested `Option`.